### PR TITLE
Implementation of moderation methods for Comments and Submissions

### PR DESCRIPTION
**Pull request by bkonyi**

### CODE REVIEW

File: lib/src/models/mixins/user_content_moderation.dart [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  /// Remove a [Comment] or [Submission].
+  ///
+  /// Set `spam` to `true` to help train the subreddit's spam filter.
+  Future remove({bool spam: false}) async => content.reddit.post(
+      apiPath['remove'], {'id': content.fullname, 'spam': spam.toString()},
+      discardResponse: true);
+
+  /// Remove distinguishing on a [Comment] or [Submission].
+  Future undistinguish() async => distinguish(how: DistinctionType.no);
```
Comment (ckartik) [inline lib/src/models/mixins/user_content_moderation.dart]: Design Question: Why did you decide to give only this enum type an exclusive no param method?

File: lib/src/models/mixins/user_content_moderation.dart [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      return 'special';
+    default:
+      throw new DRAWInternalError(
+          'Invalid user content distinction type: $type.');
+  }
+}
+
+/// Provides moderation methods for [Comment]s and [Submission]s.
+abstract class UserContentModerationMixin {
+  dynamic get content;
```
Comment (ckartik) [inline lib/src/models/mixins/user_content_moderation.dart]: Boy are mixins are cool! 😄

File: lib/src/models/submission_impl.dart [hunk 3]
```diff
@@ -382,3 +387,179 @@ class SubmissionRef extends UserContent {
...
+      'css_class': cssClass,
+      'link': _content.fullname,
+      'text': text,
+    };
+    var subreddit = _content.subreddit;
+    if (subreddit is! Subreddit) {
+      subreddit = await subreddit.populate();
+    }
+    final url = apiPath['flair']
+        .replaceAll(_subModRegExp, (subreddit as Subreddit).displayName);
```
Comment (ckartik) [inline lib/src/models/submission_impl.dart]: Why are you being forced to cast here? 

```dart
(subreddit as Subreddit).displayName
```

Guessing subreddit could either be of type `[Subreddit]` or `[SubredditRef]`, but both provide the **same** getter for the _displayName_ field.

File: lib/src/models/mixins/user_content_moderation.dart (hunk 2 above)
Comment (bkonyi) [reply lib/src/models/mixins/user_content_moderation.dart]: They're pretty useful :)

File: lib/src/models/mixins/user_content_moderation.dart (hunk 1 above)
Comment (bkonyi) [reply lib/src/models/mixins/user_content_moderation.dart]: This is just to match up with the valid values from the Reddit API.

**Pull request merged**